src/txn/txn.c

```diff
@@ -604,11 +608,21 @@ __txn_commit_timestamp_validate(WT_SESSION_IMPL *session)
 	    txn->mod_count != 0)
 		WT_RET_MSG(session, EINVAL, "no commit_timestamp required and "
 		    "timestamp set on this transaction");
+	if (F_ISSET(txn, WT_TXN_TS_DURABLE_ALWAYS) &&
+	    !F_ISSET(txn, WT_TXN_HAS_TS_DURABLE) &&
+	    txn->mod_count != 0)
+		WT_RET_MSG(session, EINVAL, "durable_timestamp required and "
+		    "none set on this transaction");
+	if (F_ISSET(txn, WT_TXN_TS_DURABLE_NEVER) &&
+	    F_ISSET(txn, WT_TXN_HAS_TS_DURABLE) &&
+	    txn->mod_count != 0)
+		WT_RET_MSG(session, EINVAL, "no durable_timestamp required and "
+		    "durable timestamp set on this transaction");
 
 	/*
 	 * If we're not doing any key consistency checking, we're done.
 	 */
-	if (!F_ISSET(txn, WT_TXN_TS_COMMIT_KEYS))
+	if (!F_ISSET(txn, WT_TXN_TS_COMMIT_KEYS | WT_TXN_TS_DURABLE_ALWAYS))
```

Comment:
This comment is a follow-on to an earlier discussion. I think you should add a `key_consistent` option for `durable_timestamp` and have this code triggered only on that. The key consistency checks are a lot more processing. The `always` and `never` settings are meant to be fast high-level sanity checks.

When you do that, you'll have to consider whether durable's key consistency checking does, or does not, imply the commit key consistency checks. Right now the code below durable does imply commit consistency checking. But key consistency checking, particularly for durable timestamps is now a very large performance hit that we might not want to pay with a simple `always` check.